sensu/goplugin.go

```diff
@@ -82,89 +102,131 @@ func (goPlugin *basePlugin) readSensuEvent() error {
-			goPlugin.cmdArgs.StringVarP(valuePtr, option.Argument, option.Shorthand, option.Env,
-				option.Default.(string), option.Usage)
-		case *uint64:
-			valuePtr, _ := option.Value.(*uint64)
-			goPlugin.cmdArgs.Uint64VarP(valuePtr, option.Argument, option.Shorthand, option.Env,
-				option.Default.(uint64), option.Usage)
-		case *bool:
-			valuePtr, _ := option.Value.(*bool)
-			goPlugin.cmdArgs.BoolVarP(valuePtr, option.Argument, option.Shorthand, option.Env,
-				option.Default.(bool), option.Usage)
+func setupFlag(cmd *cobra.Command, opt *PluginConfigOption) error {
+	viper.BindEnv(opt.Argument, opt.Env)
+	if opt.Value == nil {
+		return errors.New("nil Value")
+	}
+	if reflect.TypeOf(opt.Value).Kind() != reflect.Ptr {
+		return errors.New("Value is not a pointer")
+	}
+	value := reflect.Indirect(reflect.ValueOf(opt.Value))
+	if opt.Default != nil {
+		defaultType := reflect.TypeOf(opt.Default)
+		valueType := value.Type()
+		if t1, t2 := valueType.Kind(), defaultType.Kind(); t1 != t2 {
+			return fmt.Errorf("Value type does not match Default type: %v != %v", t1, t2)
 		}
+		viper.SetDefault(opt.Argument, opt.Default)
+	}
+	switch kind := value.Type().Kind(); kind {
+	case reflect.Bool:
+		cmd.Flags().BoolVarP(opt.Value.(*bool), opt.Argument, opt.Shorthand, viper.GetBool(opt.Argument), opt.Usage)
+	case reflect.Int:
+		cmd.Flags().IntVarP(opt.Value.(*int), opt.Argument, opt.Shorthand, viper.GetInt(opt.Argument), opt.Usage)
+	case reflect.Int32:
+		cmd.Flags().Int32VarP(opt.Value.(*int32), opt.Argument, opt.Shorthand, viper.GetInt32(opt.Argument), opt.Usage)
+	case reflect.Int64:
+		cmd.Flags().Int64VarP(opt.Value.(*int64), opt.Argument, opt.Shorthand, viper.GetInt64(opt.Argument), opt.Usage)
+	case reflect.Uint:
+		cmd.Flags().UintVarP(opt.Value.(*uint), opt.Argument, opt.Shorthand, viper.GetUint(opt.Argument), opt.Usage)
+	case reflect.Uint32:
+		cmd.Flags().Uint32VarP(opt.Value.(*uint32), opt.Argument, opt.Shorthand, viper.GetUint32(opt.Argument), opt.Usage)
+	case reflect.Uint64:
+		cmd.Flags().Uint64VarP(opt.Value.(*uint64), opt.Argument, opt.Shorthand, viper.GetUint64(opt.Argument), opt.Usage)
+	case reflect.Float32:
+		cmd.Flags().Float32VarP(opt.Value.(*float32), opt.Argument, opt.Shorthand, float32(viper.GetFloat64(opt.Argument)), opt.Usage)
+	case reflect.Float64:
+		cmd.Flags().Float64VarP(opt.Value.(*float64), opt.Argument, opt.Shorthand, viper.GetFloat64(opt.Argument), opt.Usage)
+	case reflect.Map:
+		ptr, ok := opt.Value.(*map[string]string)
+		if !ok {
+			return fmt.Errorf("only pointer to map[string]string is allowed, not %v", kind)
+		}
+		cmd.Flags().StringToStringVarP(ptr, opt.Argument, opt.Shorthand, viper.GetStringMapString(opt.Argument), opt.Usage)
+	case reflect.Slice:
+		ptr, ok := opt.Value.(*[]string)
+		if !ok {
+			return fmt.Errorf("only pointer to []string is allowed, not %v", kind)
+		}
+		cmd.Flags().StringSliceVarP(ptr, opt.Argument, opt.Shorthand, viper.GetStringSlice(opt.Argument), opt.Usage)
```

Comment:
Nice!
We probably need to add a section on how to format the input to such types; eg: is it `--flag 1,2,3` for a slice flag?

Reply:
Woof... probably